lib/config/dynamic/admiral/source.go

```diff
@@ -0,0 +1,396 @@
+	ProductVMTag       = "ProductVM"
+	admiralEndpointKey = "guestinfo.vicova.admiral.endpoint"
+
+	clusterFilter = "(address eq '%s' and customProperties.__containerHostType eq 'VCH')"
+)
+
+// #nosec
+const admiralTokenKey = "guestinfo.vicova.engine.token"
+
+var (
+	trueStr        = "true"
+	projectsFilter = "customProperties.__enableContentTrust eq 'true'"
+)
+
+// NewSource creates a new Admiral dynamic config source. sess
+// is a valid vsphere session object. vchID is a unique identifier
+// that will be used to lookup the VCH in the Admiral instance; currently
+// this is a URI to the docker endpoint in the VCH.
+func NewSource(sess *session.Session, vchID string) dynamic.Source {
+	return &source{
+		d:     &productDiscovery{},
+		sess:  sess,
+		vchID: vchID,
+	}
+}
+
+type source struct {
+	mu      sync.Mutex
+	d       discovery
+	sess    *session.Session
+	vchID   string
+	c       *client.Admiral
+	lastCfg *vchcfg.VirtualContainerHostConfigSpec
+}
+
+// Get returns the dynamic config portion from an Admiral instance. For now,
+// this is empty pending details from the Admiral team.
+func (a *source) Get(ctx context.Context) (*vchcfg.VirtualContainerHostConfigSpec, error) {
+	var err error
+	if err = a.discover(ctx); err != nil {
+		return nil, err
+	}
+
+	var projs []string
+	projs, err = a.projects(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var wl []string
+	wl, err = a.whitelist(ctx, projs)
+	if err != nil {
+		return nil, err
+	}
+
+	newCfg := &vchcfg.VirtualContainerHostConfigSpec{
+		Registry: vchcfg.Registry{RegistryWhitelist: wl},
+	}
+
+	err = a.diff(newCfg)
+	a.lastCfg = newCfg
+	return newCfg, err
+}
+
+func (a *source) diff(newCfg *vchcfg.VirtualContainerHostConfigSpec) error {
+	if a.lastCfg == nil {
+		return nil
+	}
+
+	if newCfg == nil {
+		return dynamic.ErrConfigNotModified
+	}
+
+	// compare whitelists
+	if len(a.lastCfg.RegistryWhitelist) != len(newCfg.RegistryWhitelist) {
+		return nil
+	}
+
+	for _, w1 := range a.lastCfg.RegistryWhitelist {
+		found := false
+		for _, w2 := range newCfg.RegistryWhitelist {
+			if w2 == w1 {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return nil
+		}
+	}
+
+	return dynamic.ErrConfigNotModified
+}
+
+func (a *source) discover(ctx context.Context) error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if a.c != nil {
+		return nil
+	}
+
+	token, u, err := a.d.Discover(ctx, a.sess)
+	if err != nil {
+		return err
+	}
+
+	// copied from http.DefaultTransport
+	tp := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+	cl := &http.Client{
+		Transport: tp,
+	}
+
+	if u.Scheme == "https" {
+		tp.TLSClientConfig = &tls.Config{InsecureSkipVerify: true} // #nosec
```

Comment:
@hmahmood honestly I have no idea currently. I've opened #5838 as a reminder for further work in that regard.